### PR TITLE
liveblogs share block permalinks

### PR DESCRIPTION
**Pull request by johnduffell**
at present sharing blocks uses the short urls with a hash on them.  Apparently short urls are primrily for the paper, and were used for sharing before twitter could handle long urls.  We use short urls for all sharing at the moment.

Sharing the full urls saves a redirect so makes things quicker.  Also it means we can use the full permalink in liveblog block sharing.

This PR changes it to use the full url with expanded campaign code.

I haven't tested this with all sharing, especially could do with testing on galleries sharing individual images as that uses the same block level sharing.
It would also need checking with data people to make sure the campaign codes change won't break anything.

I think this change is quite risky for causing issues with on site sharing and needs careful monitoring after go live.  As such it needs someone to take full responsibility for getting it into prod.  If it wants to go live someone should have a look at the issue starting with this PR and talking to all the necessary people outside the team to make sure nothing is missed.  Otherwise it can wait until I'm available and I'll handle it!

@stephanfowler @OliverJAsh take care!

### CODE REVIEW

**Comment by michaelwmcnamara** [comment]
Confirmed sharing campaign codes are as they currently are in live.
Tested for liveblog blocks, gallery images, images in articles and articles themselves.
Tested - for images: twitter, facebook and pinterest (nb: pinterest has no campaign code but this is the case on the current live environment, so out of the scope of this change - the other two work fine)
Tested - for liveblogs and articles: facebook, twitter, google plus and linked in (nb - linked in also has no campaign code from live)

File: common/app/model/ShareLinks.scala [hunk 1]
```diff
@@ -21,15 +22,33 @@ final case class ShareLinks(
...
+      appendQueryParams(webLinkUrl, queryParams) + elementId.map(id => s"#${id}").getOrElse("")
+    }
+
+    def appendQueryParams(uri: String, queryParams: Map[String, String]): String = {
+      queryParams.foldLeft(uri)((currentUri, queryParam) => {
+        javax.ws.rs.core.UriBuilder.fromUri(currentUri)
+          .queryParam(queryParam._1, queryParam._2)
+          .build()
+          .toString()
+      })
```
Comment (OliverJAsh) [inline common/app/model/ShareLinks.scala]: @rich-nguyen You might want to review this Scala…

Comment (rich-nguyen) [inline common/app/model/ShareLinks.scala]: add to strings implicit? its a good one!

**Comment by NataliaLKB** [comment]
:+1: :sparkles: :ship:

File: common/app/model/ShareLinks.scala (hunk 1 above)
Comment (OliverJAsh) [inline common/app/model/ShareLinks.scala]: See https://github.com/guardian/frontend/commit/3eda87061d8e2191f50c92ca3a687cdd1a5a9736

File: common/app/model/ShareLinks.scala [hunk 2]
```diff
@@ -21,15 +22,24 @@ final case class ShareLinks(
...
-    lazy val googlePlus = shareCampaignUrl("sgp", elementId).urlEncoded
-    lazy val link = shareCampaignUrl("sbl", elementId).urlEncoded
-    lazy val twitter = shareCampaignUrl("stw", elementId).urlEncoded
-    lazy val whatsapp = shareCampaignUrl("swa", elementId)
+    lazy val facebook = createShareLinkUrl(Some("sfb"), elementId).urlEncoded
+    lazy val googlePlus = createShareLinkUrl(Some("sgp"), elementId).urlEncoded
+    lazy val link = createShareLinkUrl(Some("sbl"), elementId).urlEncoded
+    lazy val twitter = createShareLinkUrl(Some("stw"), elementId).urlEncoded
+    lazy val whatsapp = createShareLinkUrl(Some("swa"), elementId)
+    lazy val pinterest = createShareLinkUrl(None, elementId).urlEncoded
```
Comment (rich-nguyen) [inline common/app/model/ShareLinks.scala]: I think this needs a campaign code too.

**Comment by rich-nguyen** [comment]
:+1:

File: common/app/model/ShareLinks.scala (hunk 2 above)
Comment (OliverJAsh) [inline common/app/model/ShareLinks.scala]: And linkedIn. Will follow up

On Wed, 3 Feb 2016 at 11:06 Richard Nguyen <notifications@github.com> wrote:

> In common/app/model/ShareLinks.scala
> <https://github.com/guardian/frontend/pull/11740#discussion_r51707151>:
>
> >      }
> >
> > -    lazy val facebook = shareCampaignUrl("sfb", elementId).urlEncoded
> > -    lazy val googlePlus = shareCampaignUrl("sgp", elementId).urlEncoded
> > -    lazy val link = shareCampaignUrl("sbl", elementId).urlEncoded
> > -    lazy val twitter = shareCampaignUrl("stw", elementId).urlEncoded
> > -    lazy val whatsapp = shareCampaignUrl("swa", elementId)
> > +    lazy val facebook = createShareLinkUrl(Some("sfb"), elementId).urlEncoded
> > +    lazy val googlePlus = createShareLinkUrl(Some("sgp"), elementId).urlEncoded
> > +    lazy val link = createShareLinkUrl(Some("sbl"), elementId).urlEncoded
> > +    lazy val twitter = createShareLinkUrl(Some("stw"), elementId).urlEncoded
> > +    lazy val whatsapp = createShareLinkUrl(Some("swa"), elementId)
> > +    lazy val pinterest = createShareLinkUrl(None, elementId).urlEncoded
>
> I think this needs a campaign code too.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/frontend/pull/11740/files#r51707151>.
>

-- 

------------------------------
This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

**Pull request merged**